lib: Make random number generation thread safe

marisn opened the pull request:
As random file generation keeps internal state in a static variable, it is prone to failures as multiple threads would attempt to update sate simultaneously. This PR adds a mutex lock around internal state update calls thus preventing clashes when called from multiple threads.

Contains code to be merged as PR #6479 
Also contains necessary changes in Make/CMake to compile gis lib with pthread support.

Should fix issues observed in r.mapcalc after thread support was merged #5742

marisn committed:
lib: correctly guard random number generation when run threaded

As random number generator tracks its internal state as a static
variable, it is necessary to guard its updates with mutex to
prevent misbehavour when multiple threads change values simultaneously

marisn commented:
> The code changes look straightforward. Do you have an idea about performance? I agree we need the change in any case, but at least knowing whether we need a better solution in the future would be good.

I haven't tested but locked part is small and should be fast (just update, no IO), thus I would expect impact on speed to be negligible.

wenzeslaus commented on lib/gis/lrand48.c:
```diff
@@ -41,22 +45,36 @@ static const uint32 b0 = 0xB;
  *
  * \param[in] seedval 32-bit integer used to seed the PRNG
+ *
+ * This function is thread-safe.
```
Something like the following should indicate that this applies only for certain builds. I don't know if you can actually have a combination when the thread safety matters while not having pthreds.

```suggestion
 * This function is thread-safe (if threading is enabled, i.e., `HAVE_PTHREAD` is defined).
```

or: ...if pthreads is available,...

marisn replied:
Disagree. This makes no sense. Of course there is no thread safety when code is compiled without threads! This is a programmer documentation that is meant to signal other developers that it is safe to call this function from threaded code. Attempts to compile core GRASS without `pthreads` and calling from a module with `pthreads` should not be considered in documentation of each function.

wenzeslaus replied:
Fair enough.

wenzeslaus replied:
pthreads and opemmp flags are independent, so strange things can happen. Now this is not about having pthreads but about having C11 compliant atomics.

wenzeslaus replied:
Since this is about having the right build with modern tooling, I leave this out of this PR.

With compliant C11, this is enabled automatically. This is now even the case for our Windows build (MinGW-w64). Current MSVC would silently fall back to the old non-thread-safe code (and continue producing bad output with multiple threads), but perhaps this will resolve itself before it is an issue for us (once MSVC's C11 atomics leave experimental status and it stops defining `__STDC_NO_ATOMICS__`, the atomic path turns on automatically), or we see whether what is available in MSVC is sufficient.

<details>

<summary>Details (AI-generated)</summary>

* The shipped Windows package (OSGeo4W) builds with MinGW-w64 GCC under MSYS2 with --with-openmp (mswindows/osgeo4w/build_osgeo4w.sh:61). GCC has C11 atomics, so the atomic path is active there — the shipped Windows build gets the fix.
* The combination that keeps the race is MSVC with OpenMP: CMake's WITH_OPENMP defaults to ON (CMakeLists.txt:149) and MSVC defines __STDC_NO_ATOMICS__, so MSVC builds fall back to the non-thread-safe code. But there is no Windows/MSVC runner in cmake.yml (ubuntu only), so "real on Windows" means the in-progress MSVC port, not any shipped or CI-tested build.

</details>

wenzeslaus commented on lib/gis/lrand48.c:
```diff
@@ -41,22 +45,36 @@ static const uint32 b0 = 0xB;
 {
     uint32 x = (uint32) * (unsigned long *)&seedval;
 
+#ifdef HAVE_PTHREAD
+    pthread_mutex_lock(&lrand48_mutex);
+#endif
+
     x2 = (uint16)HI(x);
     x1 = (uint16)LO(x);
     x0 = (uint16)0x330E;
     seeded = 1;
+
+#ifdef HAVE_PTHREAD
+    pthread_mutex_unlock(&lrand48_mutex);
+#endif
 }
```
Calling this from multiple threads will reset the global state. So, while the locking will prevent creation of a broken global state and it will prevent from reading a half-baked state, it will not work as expected. If you call this from multiple threads, they will replace the previous thread random state. Only the main thread should call this function and typically would call it once unless resetting the seed - that should be documented.

marisn replied:
You are correct. I added a note to the documentation. Mutex guards should remain in place, but seed initialization always should be called only from main process/single thread. I called multiple times, it would result in "last wins" state.

wenzeslaus replied:
The new code still requires careful setup of the state before the threads (random numbers will always require some careful setup when we want randomness, reproducibility and threads all at the same time). The new code keep this and it is document, so this is resolved.

wenzeslaus reviewed (dismissed):
I think the locking needs to expand to reading of the shared state, it is both write and read which need to be isolated.

More generally, I'm afraid we didn't finalize the _r.mapcalc_'s _rand_ behavior with multiple threads (and seeds). (Or did we?) If _r.mapcalc_ is using these functions directly as before parallelization, the random numbers placed in the raster will depend on the order of threads, i.e., what rows will come first, but I did not review the relevant code now. (An ultimate fix there would be different "context" objects for the random numbers, per row (for maximum reproducibility) or per thread (for at least seed+nprocs reproducibility).

marisn committed:
lib: ensure reproducible stream of random numbers when called from
multiple threads

As random number generation is guarded by a mutex, it makes generation
a serial operation – always leading to the same sequence of numbers

marisn commented:
> No IO, but it is called for every cell in _r.mapcalc_, no? Still better slow with overall faster code than a segfault, and `rand()` is already much slower than let's say a constant (I did not test recently). If it is slow, we could use a solution with an explicit state (context) which may allow us to avoid locking, while having a separate locking version of the API.

RN generation now is always serial – outcome order is guaranteed. Consumption is another topic. I lean towards just stating that for fully reproducible outcome one should set parameters for single thread execution.

wenzeslaus commented on lib/gis/testsuite/test_lrand48.py:
```diff
@@ -0,0 +1,108 @@
+        G_lrand48.argtypes = []
+        G_lrand48.restype = ctypes.c_long
+
+        # --- 1. Single-threaded execution ---
```
While often seen in generated code, I think our experience is that these ASCII comment graphic is hard to maintain and inherently inconsistent (thinking about old code we cleaned in GRASS).

marisn commented on lib/gis/lrand48.c:
```diff
@@ -104,6 +124,10 @@ long G_srand48_auto(void)
 
 static void G__next(void)
 {
+#ifdef HAVE_PTHREAD
+    pthread_mutex_lock(&lrand48_mutex);
+#endif
+
```
> Edit: That is if we want the locking for the states, but my original idea was that independent states don't need locking because there is nothing shared between them, and they are simply not suitable to be used across threads. In other words, a not thread-safe API meant to be used with different instances in different threads which would be then a thread-safe usage pattern.

You mean – thread-local storage? I discarded this idea because outcome is totally not reproducible. To make things worse – each thread would have to have its own unique seed value otherwise all of them would produce identical number sequences.

wenzeslaus replied:
Thread-local storage - whatever the naming, I meant passing a separate state object to each thread.

Reproducible? If each thread gets its own seed, it generates its own sequence. For a given number of threads, there is the same set numbers generated. If each thread is assigned the same rows every time it runs, the raster gets the same numbers on each position.

Each thread getting its own seed - yes, that's what I'm saying would be needed to get reproducible results. For a raster, it further depends on how threads are assigned rows, if that depends on thread's performance or it is determined ahead of time based on the number of threads (I think that's what we have in the r.mapcalc code). To make the results not only reproducible for seed+nprocs, but for seed and any nprocs number, we would have to have a separate seed for each row so that it is independent on number of threads (so each row seeded with seed plus row number).

In any case, this something extra. My point here was the missing mutex.

However, the discussion also connects to the issue of the currently generated sequence. What throws me off is your mention of discarding the idea of thread-local storage because outcome is not reproducible. In your current implementation, the number sequence is the same for every run and number of threads, but how they are distributed in practice within a raster depends on the order execution of the threads, i.e., it may differ for every run even with the same number of threads.

So, even your current implementation does not offer reproducibility. Your comment about guarantees, outcomes, and reproducibility suggests we are in agreement on that. Is that correct? I just want to be sure we are clear about what to expect from the code now.

wenzeslaus replied:
The seed generation is tested, documented, and improvements mentioned in comments. Resolving this.

echoix committed:
Merge branch 'main' into rand_thread

echoix committed:
Update lib/gis/testsuite/test_lrand48.py

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

echoix committed:
Merge branch 'main' into rand_thread

echoix committed:
Merge branch 'main' into rand_thread

echoix committed:
Merge branch 'main' into pr/marisn/6480

echoix committed:
style: Format cmake files with Gersemi

wenzeslaus commented:
Commit 38ed9dc82d updates the PR to use C11 atomics instead of a pthread mutex to fix #5742 (thread-safe `G_lrand48`/`G_mrand48`/`G_drand48`). The generated sequence for a given seed is unchanged and now pinned by the regression tests from #7633 and #7634, which pass on my fork and should run also here.

<details>
<summary>Details: why atomics, how it works, and what was verified (AI-generated)</summary>

Three problems the mutex approach ran into:

1. **The default builds never got the fix.** `HAVE_PTHREAD` is only defined with `--with-pthread`, which is independent of `--with-openmp` and off by default. Our own macOS CI and both Docker images build with `--with-openmp` but without `--with-pthread`, so exactly the builds with the #5742 race would have compiled the mutex out. The atomics gate (`__STDC_VERSION__`/`__STDC_NO_ATOMICS__`) is answered by the compiler, with no configure involvement.

2. **Build system changes are no longer needed.** The Makefile and CMake changes are dropped entirely (also resolving the CMake review comment). As a side effect this avoids defining `HAVE_PTHREAD` on MSVC, where `find_package(Threads)` succeeds with Win32 threads but `<pthread.h>` does not exist.

3. **Performance.** The mutex serializes every call and adds contention: in a microbenchmark, threaded runs were slower than single-threaded at every thread count. The atomic version keeps the exact same serialized sequence at about half the cost when several threads generate numbers at the same time, with near zero single-threaded overhead. (A single shared generator still cannot scale with threads; per-thread streams remain future work.)

**How it works:** the generator is the standard drand48 linear congruential generator (LCG): the next state is computed from the current one as X' = (A\*X + B) mod 2^48. The previous code kept those 48 bits split across three 16-bit variables and advanced them with hand-written multi-word arithmetic, a leftover from times before 64-bit integers were portable; that split is what made concurrent updates able to interleave and corrupt the state. Now the whole state is a single 48-bit value in one `atomic_uint_least64_t`, and advancing it is one atomic compare-and-swap (CAS): a thread reads the state, computes the successor with the formula above, and atomically replaces the state only if it still holds the value that was read; otherwise it retries with the fresh value. In other words, the update optimistically assumes the state was not touched by another thread in the meantime and just recomputes when it was — unlike a lock, which pessimistically excludes all other threads ahead of time and pays that cost on every call whether a conflict occurs or not. Each successful swap is exactly one generator step, and because the caller computes its return value from the successor it installed rather than by re-reading shared variables, concurrent callers consume distinct, consecutive steps of the same sequence. Without C11 atomics the code falls back to the previous plain implementation (not thread-safe, i.e. unchanged behavior on such platforms; RFC 7 allows optional C11 features with a fallback).

The same C11-atomics-with-fallback approach is used in #7350; if both land, the capability test can be factored into a shared macro.

**Reproducibility:** the sequence of values is serialized and seed-reproducible, but which thread (and so which raster cell) receives which value depends on scheduling. Fully reproducible results still require single-threaded execution; this is now stated in the file documentation. The concern about `G_fatal_error()` under a held lock in the mutex-based implementation no longer applies (there is no lock).

**Test cleanup:** the duplicate-values assertion is removed (a correct random sequence can contain duplicates, as the observed `10000 != 9996` failure showed; the sorted-lists comparison is the valid check), the AI authorship note moved to the PR, and the banner comments are gone.

The 64-bit reformulation and its verification were AI-assisted (Claude). Verification included a standalone build of this file compared against the previous implementation for 100k steps and the reference sequences for five seeds, an 8-thread multiset check (1M draws), and the regression tests from #7633 and #7634 run against a GRASS build using this file.

</details>

wenzeslaus commented on lib/gis/testsuite/test_lrand48.py:
```diff
@@ -0,0 +1,108 @@
+
+        G_lrand48.argtypes = []
+        G_lrand48.restype = ctypes.c_long
+
```
Seems that this is done (by ctypesgen) and the code runs in CI.

wenzeslaus commented on lib/gis/CMakeLists.txt:
```diff
@@ -4,7 +4,13 @@ if(NOT WIN32)
+if(Threads_FOUND)
+  list(APPEND grass_gis_DEFS "HAVE_PTHREAD")
+endif()
+
```
This now uses existing C11 macro, so no changes to the build system.

wenzeslaus commented:
@nilason This is now using your approach from #7350 while not touching any build files (unlike the pthreads locking version). Please review.

wenzeslaus reviewed (commented):
This is passing tests from #7633 and #7634, i.e. generating same numbers as before. I approve, but I'm also responsible for the new code, so someone else should give the formal approval, too.
